### ISSUE TITLE
[Enhancement] Add appsettings support

### ISSUE THREAD

**Issue by Redth** [dotnet/maui#4408]
### Description

It would be useful to have a way for apps to build with different settings (ie: the `appsettings.json` file pattern already common with Microsoft.Extensions.*).

This is often done like in: https://github.com/dotnet/maui/pull/3872 

There's discussion around removing hosting for .NET 6 MAUI GA: https://github.com/dotnet/maui/issues/4393

In the future we should find a path forward to support this.

One idea is to use a source generator to create the code to inject values from the .json files at build time so that we do not pay a performance price for reading and parsing json files from embedded resources.

**Comment by hflexgrig** [comment]
> One idea is to use a source generator to create the code to inject values from the .json files at build time so that we do not pay a performance price for reading and parsing json files from embedded resources.

@Redth Thats gonna scaffold C# classes from json file on build time? or gonna work like IL code injection?

**Comment by maxkoshevoi** [comment]
Also support for user secrets would be great. Currently have to work around it like this: https://github.com/maxkoshevoi/NureTimetable/blob/a67e7ed78cfc4de3512d8c36ac33fdefa48a8571/NureTimetable.Core/Models/Consts/Keys.cs

**Comment by maxkoshevoi** [comment]
Related / duplicate of https://github.com/dotnet/maui/issues/3446

**Comment by jamesmontemagno** [comment]
Is possible today via -> https://montemagno.com/dotnet-maui-appsettings-json-configuration/ however some startup performance considerations with this method.

**Comment by edgiardina** [comment]
Upvote. Supporting native appsettings files, plus support for user secrets management, prevents me from having to constantly ignore the appsettings.json file in git.

**Comment by Ghevi** [comment]
I don't know if it's already included, but can you also consider supporting `IOptionsMonitor/IOptionsSnapshot`? 
I have a scenario where I need to reload settings at runtime, I understand it's something not many people have to do, but it's possible it would be awesome.
Is there a way to do this at the moment? 
The best way i found is:
```
   configuration["MyOptions "] = JsonSerializer.Serializer(new MyOptions { value = "new value" });
   (configuration as IConfigurationRoot).Reload();
```
and subscribe with the reload token of `IConfiguration` in the services where i need those options.

**Comment by thomasbach-dk** [comment]
... and two years later there still don´t seems to be any official solution for this?

**Comment by thomasbach-dk** [comment]
> What is the use case of having appsettings? One can't easily apply different settings per device which negates the need for a configuration system.
> 
> 
> 
> `Microsoft.Extensions.Configuration` also has the negative effect of delaying startup due to dll loading and appsettings deserialization, and possibly waiting for settings to arrive before initializing the app (depending on the app).

From my point of view I need an option to supply settings (e.g. ApiUrl, Sentry DSN, OIDC Authority etc.) at build time. I can replace settings in appsettings.json or by environment variables in our GitHub Actions Workflow, and thereby avoid hard coded settings.

**Comment by espenrl** [comment]
I'll provide an alternative way to deal with settings in an app for anyone interested.

Since there are no secrets in an app package all values might as well be hardcoded - even in source control. I have TEST and PROD environments and use compiler conditional constants to control my hard coded values.

Trigger MSBuild
``` cmd
dotnet publish /p:MyEnvironment=ENVIRONMENT_TEST
```

Csproj
``` xml
<DefineConstants>$(DefineConstants);$(MyEnvironment)</DefineConstants>
```

``` csharp
#if ENVIRONMENT_TEST
  var url = "";
#else
  var url = "";
#endif
```

**Comment by thomasbach-dk** [comment]
> I'll provide an alternative way to deal with settings in an app for anyone interested.
> 
> Since there are no secrets in an app package all values might as well be hardcoded - even in source control. I have TEST and PROD environments and use compiler conditional constants to control my hard coded values.
> 
> Trigger MSBuild
> 
> ```batchfile
> dotnet publish /p:MyEnvironment=ENVIRONMENT_TEST
> ```
> 
> Csproj
> 
> ```
> <DefineConstants>$(DefineConstants);$(MyEnvironment)</DefineConstants>
> ```
> 
> ```cs
> #if ENVIRONMENT_TEST
>   var url = "";
> #else
>   var url = "";
> #endif
> ```

That´s actually how we´re handling it now. And you're right - it´s probably ok having these kinds of settings hardcoded and checked in. But imho there should be some official best practices about this related to MAUI :)

**Comment by peterblazejewicz** [comment]
> What is the use case of having appsettings?

Being in a large company that's migrating all existing old school (WinForm, WPF, UWP, WinUI, Xamarin, did I listed 5 co-existing tech?) into one dev model based on Maui, starting in the desktop app context. Settings files serve well as container for build-time configuration/transformation (i.e. in the build context of AZ pipelines). Mobile apps are only part of .Maui now IMO

**Comment by infofromca** [comment]
Another situation to use appsettings.json 👍 https://learn.microsoft.com/en-us/dotnet/core/resilience/http-resilience?tabs=dotnet-cli
```
{
    "RetryOptions": {
        "Retry": {
            "BackoffType": "Linear",
            "UseJitter": false,
            "MaxRetryAttempts": 7
        }
    }
}
```

**Comment by kubaflo** [comment]
Closing as a duplicate of #3446, which already tracked the same appsettings.json support request. If there is a remaining gap on current MAUI, please open a fresh focused issue that references the canonical thread.

**Comment by ingenobajon** [comment]
> Closing as a duplicate of [#3446](https://github.com/dotnet/maui/issues/3446), which already tracked the same appsettings.json support request. If there is a remaining gap on current MAUI, please open a fresh focused issue that references the canonical thread.

That's unfortunate. The remaining gap is precisely the lack of `appsettings.json` support described above, which is still not available in current MAUI.

**Comment by kubaflo** [comment]
Reopening — the duplicate closure was too aggressive.

#3446 asked for automatic runtime loading of `appsettings.json` through the original hosting-based configuration flow and was closed as `not planned` in 2022. This proposal was filed later by a MAUI team member in the context of removing hosting and explores a broader/different path: build-time configuration, source generation to avoid startup cost, user-secrets patterns, and related options support.

It also intentionally remained open for four years after #3446 was closed. The continuing lack of first-class appsettings support is the subject of this proposal, not a residual gap that should be moved to a third issue.

Sorry for the mistaken closure. Leaving this open for an explicit product decision.

> Disclosure: this comment was generated by AI.

**Comment by felipebazz** [comment]
What kind of evidence is needed to reinforce the importance of an `appSettings` feature?

What about the case of an app from a national company that handles financial transactions that has been using a "quick-and-dirty" `appSettings` workaround for two years and running smoothly?

You can roll out this feature quietly; we won't mind. You can even make an AI-related announcement the headline, the important thing is simply having this crucial feature officially available.
